File: javascript/ch1/ex08.js
var cbrt = function(){
  var cube = function(x) { return x * x * x}
  var abs = function(n) { return n < 0 ? n * -1 : n }
  var improve = function(x, y) { return ((x/(y*y)) + (y * 2))/3.0 }
  var isGoodEnough = function(guess, prevGuess, x) { return abs(((cube(guess) - cube(prevGuess))/guess)) < 0.001 }
  return function(x, guess, prevGuess) {
    if(!guess) { guess = 1.0 }
    if(!prevGuess) { prevGuess = 0.0 }
    if(isGoodEnough(guess, prevGuess, x)) {
      return guess;
    } else {
      return arguments.callee(x, improve(x, guess), guess);
    }
  }
}()

File: clojure/ch1/ex08.clj
(ns sicp.ch1.ex08)

(defn- cube [x] (* x x x))

(defn- abs [n] (if (< n 0) (* n -1) n))

(defn- improve [x y] 
  (/ (+ (/ x (* y y)) (* 2 y)) 3))

(defn- good-enough? [guess prev-guess x]
  (< (abs (/ (- (cube guess) (cube prev-guess)) guess)) 0.001))

(defn cbrt
  ([x] (cbrt 1.0 0.0 x))
  ([guess prev-guess x] 
     (if (good-enough? guess prev-guess x)
      guess
      (cbrt (improve x guess) guess
	   x))))
